Обновление конфига

ssysoenko opened the pull request:
Попытка сделать код читабильнее сохранив идею.
Информация для распарса сознательно теперь не в типах (создаются `FieldDescriptor` и сохраняются в контейнер, есть виртуальные вызовы и т.д.). Сделано это чтобы описание конфига выглядело как описание структуры.
При объявлении струкутуры конфига обязательно указывать описание как всего конфига так и всех его полей. Конфиг умеет выдавать свое описание.

Нововедения:
1. Описание струкутур
2. Значение возвращается в `boost::outcome`

Из хорошего пропало:
1. Создание струкруты после извлечения значений. Теперь струкутура создается сразу и значения при распарсе проставляются сразу в нее
2. Информация не только в типах. Дескрипторы полей конфига переопределяют несколько методов, для каждого поля создается экземпляр дескриптора, есть контейнер с указателями на сескрипторы.

shkryab committed:
Add new config src

Remove prev config
Add new config's src
Update test

roman-orekhov commented on include/cvs/common/config.hpp:
```diff
@@ -1,57 +1,207 @@
+    BaseFieldDescriptor() { Self::descriptors().push_back(this); }
+    virtual ~BaseFieldDescriptor() = default;
+
+    virtual bool has_dafault() const = 0;
```
has_d**e**fault

roman-orekhov commented on include/cvs/common/config.hpp:
```diff
@@ -1,57 +1,207 @@
+    }
+
+    std::string describe(std::string_view prefix) const override {
+      return prefix.data() + fmt::format(BaseFieldDescriptor::description_format, field_name_str, field_base_type_str,
```
Не эффективно (хотя важно ли это тут?). Можно в description_format спереди добавить {}, и в него уже пихать prefix.

roman-orekhov commented on include/cvs/common/config.hpp:
```diff
@@ -1,57 +1,207 @@
+        if (auto iter = ptree.find(field_name_str); iter != ptree.not_found()) {
+          config.*pointer = T::make(iter->second).value();
+        } else
+          config.*pointer = T::make(boost::property_tree::ptree{}).value();
```
То есть дефолтные значения для типа предопределены? Для чисел -- ноль, для строк -- пустая...

ssysoenko replied:
Выше проверяются вложенные поля. Если все поля с типами default и optional - то в конфиге может отсутствовать кусок для этой вложенной струкутуры и ее можно проинитить пустым деревом.

roman-orekhov replied:
Я считаю нужным дефолт для вложенного определять индивидуально. Пример в Азимуте был: была структура, определяющая параметры для сохранения картинок в ImageSaver. По дефолту выбрано было jpg с параметрами в виде массива [1,100] (так в opencv кодируется 100% качество в jpg). Ясное дело, что для другого случая имеет смысл другой дефолт (png с другим массивом, который определяет степень сжатия).

ssysoenko replied:
Надеюсь правильно тебя понял: ты пишешь о том, чтобы дефолты для вложенных струкутур определять так же как и для обычных полей:
`CVS_FIELD_DEFAULT(nested, NestedConfig, {some_values, bla, bla, bla},  "Comment");`
?
В том виде как написано сейчас такое, похоже, не получится. Надо возвращаться к старому варианту где поля перечисляются внутри макроса. Когда поля  - аргументы макроса мне не нравится эстетически.

roman-orekhov replied:
Не обязательно так, можно требовать создать переменную вне макроса, и уже её передавать по имени. В конце концов, ровно этот дефолт может не в одном месте быть нужен.

efimov-evgeniy replied:
А дефолты для отдельных полей и для всей структуры как будут взаимодействовать?

roman-orekhov commented on src/config.cpp:
```diff
@@ -2,92 +2,20 @@
+boost::property_tree::ptree CVSConfigBase::load(const std::string& content) {
   std::stringstream ss;
-  ss << file_content;
+  ss << content;
```
Есть конструктор для этого.

roman-orekhov reviewed (approved):
Печаль, конечно, что очередной конфиг пишется на ptree. Хотелось бы уже какую-нибудь либу, умеющую различать true от "true". (Кстати, полезно иметь тест на bool, да и на string, раз уж на то пошло). Иная либа также могла бы помочь с сериализацией (полезно для передачи заказчику дефолтных настроек).

В целом для решения на макросах выглядит норм :)
Отмечу только, что слишком много сложений строк для проекта, где уже используется fmtlib, и можно использовать fmt::format и fmt::join везде.
Также я бы сократил `params.value()` до `params()`, а то слишком неудобно пользоваться будет.

ssysoenko commented:
> Печаль, конечно, что очередной конфиг пишется на ptree. Хотелось бы уже какую-нибудь либу, умеющую различать true от "true". 

За буст то, что это буст и все его знают и всюду он есть. В части json мне нравится [эта](https://github.com/nlohmann/json) либа. Но нравится она мне исключительно по описанию. Использовать ее мне не довелось. Думаю загнать под условную компиляцию разынуе либы для json.

> (Кстати, полезно иметь тест на bool, да и на string, раз уж на то пошло). Иная либа также могла бы помочь с сериализацией (полезно для передачи заказчику дефолтных настроек).
 
 Учту.
 
> Отмечу только, что слишком много сложений строк для проекта, где уже используется fmtlib, и можно использовать fmt::format и fmt::join везде.

Ну это я просто утомился и поспешил выложить. :)

> Также я бы сократил `params.value()` до `params()`, а то слишком неудобно пользоваться будет.

Тут та же история, что и с json - уже есть буст и чо дергаться. Альтернатива [вот](https://github.com/TartanLlama/expected). Ну а вообще вроде подобное должно появиться в std.

efimov-evgeniy commented on include/cvs/common/config.hpp:
```diff
@@ -1,57 +1,207 @@
+
+#define CVS_CONFIG(name, description)                            \
+  static constexpr const char* name##_name        = #name;       \
+  static constexpr const char* name##_description = description; \
```
А зачем тут глобальные переменные? Передай имя как аргумент шаблона

ssysoenko replied:
Я не проверял и прибывал в уверенности, что строковый литерал как он есть в шаблон не влазит.

efimov-evgeniy replied:
Если не лезть в с++2a, можно вот так: https://github.com/Centimo/Meta_lib/blob/master/Constexpr_string.hpp

ssysoenko replied:
Я смотрю примеры применения и всюду нельзя напрямую в шаблон строковый литерал запихать.

efimov-evgeniy replied:
Если использовать код по ссылке, то делается так:
`SomeClass<decltype(get_constexpr_string([]() {return "literal"; }))>`
Соответственно, `decltype(get_constexpr_string([]() {return "literal"; }))` можно запихнуть в макрос.

roman-orekhov replied:
Ну то есть снова макрос будет, и тогда в чём разница?

efimov-evgeniy replied:
Разница с чем? С тем, что сейчас есть в этом ПР в том, что не будет лишних сущностей в глобальной области видимости.

efimov-evgeniy commented on include/cvs/common/config.hpp:
```diff
@@ -1,57 +1,207 @@
+  };
+
+  // Field with default value
+  template <typename T,
```
Есть смысл делать отдельный класс под дефолтное значение? Можно же просто в шаблон булеву переменную прокинуть. Или сделать перечисление и соединить ещё и optional

ssysoenko replied:
Да. Именно так переписать и нужно.

efimov-evgeniy commented on test/configtest.cpp:
```diff
@@ -1,170 +1,63 @@
+  CVS_FIELD_OPT(value2, float, "Nested 0. Optional field 2");
+};
+
+CVS_CONFIG(TestConfig, "Test config") {
```
Фигурные скобки так важны? Без них можно было бы сделать всё на типах.
Возможно, получится сделать и с фигурными. Я даже примерно знаю как. Или нет...

ssysoenko replied:
Тут сознательно пожертвовано правильными вещами (дескрипторы только типами без выделения памяти, сначала распарс потом создание и инициализация структуры конфига) чтобы выглядело краше (правда это субъективщина). Но вроде как не все так плохо: большого перерасхода памяти нет (пара указателей +/-), виртуальные взовы тоже на производительность не влияют, ну сами струкутуры не имеют сложной инициализации поэтому парсинг -> создание и инициализация - оно каэшн по книжке и правильно, но в данном случае ничего не дает.
В целом да, можно только на типах, но без `{}`.

efimov-evgeniy replied:
Просто фигурные скобки мне кажутся не особо полезными (и даже чуть-чуть вредными), а вот ошибки на этапе компиляции я очень люблю

ssysoenko replied:
> фигурные скобки мне кажутся не особо полезными (и даже чуть-чуть вредными)

Почему вредные?

> а вот ошибки на этапе компиляции я очень люблю

дык строгая типизация никуда не делась. Просто теперь не тайпдеф на тапл описывает все поля, а есть константы с дескрипторами и вектор указателей на эти константы.

efimov-evgeniy reviewed (changes requested):
Хотелось бы использовать типы

ssysoenko commented:
> Хотелось бы использовать типы

Зачем?

ssysoenko committed:
Minor fix

shkryab committed:
Template specialization for FieldDescriptor

Add optional nested field

shkryab committed:
Config name and description as template args

shkryab committed:
Fields name, description and type as template args

roman-orekhov commented on test/configtest.cpp:
```diff
@@ -24,11 +24,18 @@ CVS_CONFIG(TestConfig, "Test config") {
+  CVS_CONFIG(NestedConfig2, "Nested config 2") {
+    CVS_FIELD(value0, float, "Nested 2. Default field 0");
+    CVS_FIELD(value1, float, "Nested 2. Optional field 1");
+  };
```
Текст не соответствует макросам

roman-orekhov commented on include/cvs/common/config.hpp:
```diff
@@ -264,7 +278,8 @@ struct CVSConfig : public CVSConfigBase {
-  struct name : public cvs::common::CVSConfig<name, name##_name, name##_description>
+#define CVS_CONFIG(name, description)                                                              \
+  static constexpr const char* name##_name        = #name;                                         \
+  static constexpr const char* name##_description = description;                                   \
```
Зачем это оставлено, если ты-таки решил заюзать constexprString? Казалось бы только для целей избавиться от мусорных констант

ssysoenko committed:
Fix

Remove unnecessary constatnts
Add macro to simplify ConstexprString type declaration
Fix test config description

ssysoenko committed:
Backward compatibility

ssysoenko committed:
Fix backward compatibility

Add install files
Fix default value
Add opertor* and opertor* to CVSOutcome

ssysoenko committed:
Add vector and optional vector
